### ISSUE TITLE
[RFC] Proposal For Configuration Section Refactor

### ISSUE THREAD

**Issue by jeremymeyers** [stashapp/stash#1549]
*### Scope*
Rework the config page for better organization.

##  Long Form
As  #1548 is moving scrapers to its own page, I think it would be good to take a look at the configuration section overall.

My suggestions are as follows:

# Separate pages into collapsible sections, expanded by default 
(like this: https://www.codehim.com/bootstrap/bootstrap-vertical-menu-with-submenu-on-click/ )
![Screenshot 2021-07-02 16 57 22](https://user-images.githubusercontent.com/3514095/124327906-a3fefe80-db56-11eb-9277-d6a288870716.png)

# Move stuff around
(in no particular order)

## Content Section
The purpose of the Library section is to show configuration settings related to WHICH CONTENT IS INDEXED OR NOT INDEXED
### Library
- Video Extensions
- Image Extensions
- Gallery Zip Extensions
- Excluded Video Patterns
- Excluded Image/Gallery Patterns
### Tasks
- Auto Tag (Rename "Detect Tags From Filenames")
- Scan (Rename "Add To Library....").  Also rename "Scan" and "Scan Selected" to "Scan All" and "Scan Folders..."
- Generated Content (Rename "Generate Interface Files....")

## System Section
The purpose of the System section is to show configuation options and settings related to BACK-END SETTINGS
### Paths Menu
- Database Path
- Generated Path
- Cache Path

### Logs Menu
- Log File
- Log Level
- Current contents of "Log" page

### Maintenance Menu
- Clean Database
- Backup
- Metadata Export

## Indexing Section
The purpose of the System section is to show configuation options and settings related to THE INGESTION OF CONTENT
- Hashing
- Parallel Scan/Generation
- Preview Generation

## Playback Section
The purpose of the Playback section is to show configuration options and settings related to VIDEO PLAYBACK
### Playback Menu
- Auto-Start Video
- Maximum Loop Duration

### Conversions Menu
- Transcoding options
### DLNA Menu
- Existing DLNA Menu
(Possible future additions: Global playback speed, video player selection, etc)

## Users Section
The purpose of the Users section is to show configuration options and settings related to USER LOGIN
 - (contents of Authentication section)
 - Handy Connection Key

## Appearance Section
The purpose of the APPEARANCE section is to show configuration options and settings related to USER INTERFACE
### Interface Menu
- Language
- Menu Items (Rename "Enable/Disable sections")
- Wall Settings
- Slideshow Delay
### Theme Menu
- Page showing "Currently Active Theme" with source and link to themes index on wiki (or native display of that page)

### Plugins Manu
The purpose of the Plugins section is to show configuration options and settings generated by PLUGINS
- "Installed Plugins", current "plugin" page but refactored as a table with Name, Description, Version, Link. 
- "Refresh Plugin List" to replace "Reload Plugins" (at some point in the future, an "Add plugins" page)
### Scrapers Menu
(as described in #1548 )
### Tools Menu
- Existing "Tools" section

**Comment by echo6ix** [comment]
I think you have the settings page organized well.

### Settings Layout

My only suggestion here would be to rename some headings to make them more specific, and some of these might seem pedantic and subjective so I apologize in advance. For example,

- Rename anything that is called `Paths` to something more specific (and this applies to any heading that is not unique), such as:
  - `Application Paths` for application system path locations
  - `Library Paths` for library path locations
- Rename `Display` section to `Appearance` or `Interface`. In my experience, "Display" often pertains to display monitor settings, and I've noticed apps use headings like "Appearance" or "Interface" to denote the GUI.
- Rename `Themes` back to `Custom CSS` as it is now. Not everyone uses community made CSS that some people refer to as "themes"; for instance I only apply a handful of CSS tweaks to refine the stock UI myself.

### Settings GUI

I think everything in the settings page should be filterable with a [filter-on-search type mechanism](https://www.w3schools.com/howto/howto_js_filter_lists.asp) that a lot of settings pages have (Chromium, Visual Studio Code, etc). And that's part of the reason why I suggested to make headings more specific.

I tend to favour a Material Design ui, like the new Microsoft Edge settings page (and to a lesser degree the Chromium or Chrome settings page)

Desktop layout
![Screenshot 2021-07-09 001603](https://user-images.githubusercontent.com/37937507/125022725-52211180-e04b-11eb-8eb7-beada7cff08c.jpg)

Mobile layout
![Screenshot 2021-07-09 001901](https://user-images.githubusercontent.com/37937507/125022751-606f2d80-e04b-11eb-9bb7-4726dba22eac.jpg)

**Comment by jeremymeyers** [comment]
@WithoutPants thoughts on this? I can give it a whirl if you like.

**Comment by WithoutPants** [comment]
That looks fine. I don't think there's any support for theming beyond setting the custom css content.

**Comment by echo6ix** [comment]
I've done my best to organize the structure of the settings page. Please contribute any modifications or suggestions.

# Settings structure
My settings tree will use the follow structure:

- **PRIMARY HEADING TITLE**
  - Sub-heading title
    * Item belonging to the sub-heading `<form control item>`

## Primary headings
- Primary headings are displayed on a left column.
- The left column uses an absolute position; that is, it does not scroll with the page, it is static and always viewable
- Similar to the Chromium based browser settings pages, primary headings should be displayed on the main background color
- Primary headings navigate to all child sub-headings on the right column of the page

## Sub-headings
- Sub-headings are displayed on the right column.
- Sub-heading titles are displayed above a card (not inside it)
- Sub-heading descriptions are displayed beneath a sub-heading title in smaller font, and also above a card (not inside it)

## Sub-heading items
- Items are displayed within the card
- Each item is seperated by a full-bleed (edge of card to edge of card) horizonal line
- Item headings are displayed left justified and bold
- Item descriptions are placed beneath item headings and are displayed in normal weight text
- Item form elements (text input, toggle switches, buttons, etc) are displayed right justified to the item headings and description, unless it is a wide text input then it should probably be placed below the item title and description
- Use toggle switches instead of checkboxes

## Misc
- The width of the settings page should be increased 

# Settings tree
- **GENERAL**
  - Library paths
  - Application paths
    * Database Path `<text input form>`
    * Generated Path `<text input form>`
    * Metadata Path `<text input form>`
    * Cache Path `<text input form>`
    * Custom Performer Image path `<text input form>`
    * Etc.
  - Media content extensions
    * Video Extensions `<text input form>`
    * Image Extensions `<text input form>`
    * Gallery Zip Extensions `<text input form>`
  - Media exceptions
      * Excluded video patterns `<text input form>`
      * Excluded Image/Gallery patterns `<text input form>`
  - Gallery and image options
    * Create galleries from folders containing images `<toggle switch>`
    * Write image thumbnails `<toggle switch>`
  - Hashing
    * Calculate MD5 for Videos `<toggle switch>`
    * Generated file naming hash `<drop down selection>`
  - Transcoding
    * Maximum transcode size `<drop down selection>`
    * Maxium streaming transcode size `<drop down selection>`
  - Parallel Scan/Generation
    * Number of parrallel task for scan/gernation `<text input form>`
  - Preview Generation
    * Preview encoding present `<drop down selection>`
    * Include audio `<toggls switch>`
    * Number of segments in preview `<text input form>`
    * Preview segment duration `<text input form>`
    * Exclude start time `<text input form>`
    * Exclude end time `<text input form>`
- **INTERFACE**
  - Basic settings
    * Language `<drop down selection>`
    * Open web browser during startup `<toggle switch>`
    * Handy connection key `<input text form>`
    * Funscript offset `<input text form>`
    * Slideshow delay on galleries `<text input form>`
  - Customize menu items
    "Select which menu items to display on the navigation bar"
    * Show scenes `<toggle switch>`
    * Show images `<toggle switch>`
    * Show movies `<toggle switch>`
    * Etc
  - Wall view (for scenes and markers)
    * Show title and tags `<toggle switch>`
    * Enable sound `<toggle switch>`
    * Preview type `<drop down selection>`
  - Grid view (for scenes)
    * Show studios as text `<toggle switch>`
    * Auto-start video `<toggle switch>`
    * Auto-start video when playing selected `<toggle switch>`
    * Continue playlist by default `<toggle switch>`
    * Maximum loop duration `<text input form>`
  - Editing
    * Disable drowndop create
      * Performer `<toggle switch>`
      * Studio `<toggle switch>`
      * Tag `<toggle switch>`
  - Custom CSS injection
    * Enable custom CSS `<toggle switch>`
    * Custom CSS `<multiline input form>`
  - Delete Settings
    "Default settings when deleting content"
    * Delete files by default `<toggle switch>`
    * Delete generated supporting files `<toggle switch>`
- **TASKS**
    - Library tasks
        * Scan library
        * Identify
        * Auto Tag
        * Clean
    - Generated Content
        * Generate
    - Metadata
        * Full export
        * Full import
        * Import from file
    - Backup
        * BAckup
        * Download Backup
    - Migrations
        * Rename generated files
- **EVENT LOGS**
  - Log Viewer
  - Log Settings
    * Log File
    * Log to terminal `<toggle switch>`
    * Log level `<drop down selection>`
    * Log http access `<toggle switch>`
- **PLUGINS**
- **SECURITY**
  - Authentication
    * Username `<text input form>`
    * Password `<text input form>`
    * API Key `<text input form>`
    * Maximum Session Age `<text input form>`
    * Trusted Proxies
- **SERVICES**
  - DNLA
    * Status
    * Actions
    * Settings
  - Stash-Box Integration
    * Stash-box endpoints, etc
- **WEB SCRAPING**
  - Scraper settings
    * User agent `<text input form>`
    * Chrome CDP path `<text form>`
    * Check for insecure certificates `<toggle switch>`
    * Excluded tag patterns `<text input forms>`
  - Scrapers
    "Explain what scrapers are with link to community scrapers"
    * Reload scrapers `<button>`
    * Scene scrapers `<chevron up/down button>`
    * Gallery scrapers `<chevron up/down button>`
    * Etc
- **ABOUT**
  * Stash logo and info
  * Latest version build hash: xxxx `<check for new version button>`

# Notes
- I wasn't sure if the WALL VIEW settings applied to all wall views or just scenes and markers, hence why I specified scenes and markers in parantheses
- Ditto for GRID VIEW settings

**Comment by WithoutPants** [comment]
From Fonzie on Discord:

> We don't need to split up Settings and Tasks, and we also don't need a 3-level deep Settings tree where you spit up the two desktop page-long Interface in 7 different sub-trees so that "Editing" with its current 6 lines ends up on its own single page. Using a sledgehammer to crack a nut won't make things better.
> 
> The sheer amount of information on these pages (with Configuration page maybe as the exception) isn't the problem or that the current 9 nav-items are too many. 
> The two big problems are the spacing between the sections/form-groups within the pages so you don't really get a sense when one section ends and a new begins, and that everything is too vertical cause the design does a poor job of using the width of the page. 
> (Plus #3: quite a few options are poorly explained.)
> 
> From an early survey result we know that a large majority mostly access Stash from Desktop and only 5% Mobile Only. So I can't figure out why seemingly most design ideas are mobile first desktop second, and/or try to copy a solution from & for browser settings rather than one from & for website settings.

![image](https://user-images.githubusercontent.com/53250216/143505850-b4c0453e-d0e1-4168-b2ed-1cc67de16da5.png)

**Comment by echo6ix** [comment]
> we also don't need a 3-level deep Settings tree...

Is that referencing the settings tree in this thread because I don't understand how the criticism is applicable.

If so, I'm not sure the commenter understands the purpose of the settings tree is just a semantical representation of all the settings to help us organize them. I was going to write it in YAML but I thought it would be clearer for most people the other way. If we're going to refactor the settings, we need to compile a list of the main categories and most of the items, so we can decide if we won't to consolidate, split, purge, or shuffle sections, and at least the settings tree gives us some a basic reference point.

It's not a literal tree UI, and I guess I may not have been clear enough in the structure section where it defines primary headings are left column, sub-heading cards are right column, and the items are contained within the sub-heading cards. The images he references use the exact same "three level deep" structure I outlined:

```
<h1>Preferences</h1> <----------- Primary heading , displayed on the right, "level 1"

<h2>About Preferences<h2> <----------- Sub-headingc "level 2"
<p>Item 1</p> <----------- Item, displayed on the right, "level 3"
<p>Item 2..</p> <----------- Item, displayed on the right, "level 3"

<h2>People You Follow<h2> <----------- Sub-heading, displayed on the right, "level 2"
<p>Item 1</p> <----------- Item, displayed on the right, "level 3"
<p>Item 2..</p> <----------- Item, displayed on the right, "level 3"

```
Gonna be brutally honest. Those pics as a guideline are hideous. I don't know any modern app with simple and intuitive UI/UX that is using an archaic message board Web 1.0 layout as its inspiration. That is literally a step backwards, like 15 years in the past backwards. There is a reason iOS, Google's Material (Chromium, Android, etc), Windows 11 settings, some Linux DEs, are all using a "three level deep" card layout for settings.

The card layout solves the "two big problems" precisely -- since each sub-heading group lives on it's own island in a card. i don't think another visual layout exists that conveys to the user "this group of items here on this card are all categorically related to one another and separate from the other sections that live on the other cards", so I'm confused at the criticism or maybe just misunderstanding what it is directed at.

I'll see if I can build a wireframe later, and I'll edit my prior post to make it clear that the settings tree is just a semantical reference.

**Comment by 7dJx1qP** [comment]
I think "System" is a better name for the "General" heading.

I would also move it down on the list. I don't think any of the settings in that section are modified that frequently, if at all. Library paths are most likely to change, but I think it's something that happens occasionally if at all. If someone setting up stash for the first time sets their paths in the wizard, then I can see them not ever needing to even go to the system section.

If we look at browser settings, they even hide the System section under a collapsible "Additional settings" header. I don't think we need to go that far, but I don't think our equivalent of system settings should be first on the menu. It's not even the default tab if you just go to http://localhost:9999/settings. The default is Tasks, so it would make more sense to put that at the top.

**Comment by echo6ix** [comment]
@7dJx1qP All good suggestions

Here's a wireframe of what I'm talking about. Ignore the lack of items or how they're categorized, I just threw in a few to portray the gist of the UI.

Image attached below or [clickable in a different window here](https://user-images.githubusercontent.com/37937507/143616769-028a1262-0191-420f-8e78-e3411dd445a2.png)

![by6JTWovaLh3ccEuHmvCBn](https://user-images.githubusercontent.com/37937507/143616769-028a1262-0191-420f-8e78-e3411dd445a2.png)

**Comment by WithoutPants** [comment]
I think `General` is better renamed to `Library`, and that the `Application Paths` should be moved to a different panel (ie System). Application paths are not something that are likely to be touched very often, and mostly for exceptional circumstances. As mentioned above, I'd make `Tasks` first, then `Library`, then the rest with `System` near the bottom.

I don't think stash-box integration belongs in Services. Services implies stuff provided by the stash server (ie DLNA). I would say it better belongs in Web Scraping, though that title probably needs generalising as a result.

I like the wireframe posted above. My only addition there is that it needs to be clear in the interface when there are unsaved changes to the configuration. A floating save button at a bear minimum along with a warning if navigating away with unsaved changes.

**Comment by echo6ix** [comment]
> My only addition there is that it needs to be clear in the interface when there are unsaved changes to the configuration. A floating save button at a bear minimum along with a warning if navigating away with unsaved changes.

@WithoutPants Just curious, why can't user input to all the form items be automatically saved in the background [[source](https://www.chevtek.io/how-to-implement-autosave-in-your-web-app/)]. This would nullify the purpose of a save button.

**Comment by echo6ix** [comment]
> Just curious, why can't user input to all the form items be automatically saved in the background
Upon some thought, something like this would need to be disabled for the CSS section. In fact, I have given some thought to the layout `Custom CSS` subsection.

### Custom CSS subsection
To reduce clutter on the main settings page and to increase the usability of the Custom CSS input box, herein referred to as the `CSS Editor`, add an external launch icon that opens a new browser window specifically for editing custom CSS.

#### CSS Editor
- Dedicated save button
- Line numbers, alternate line colors, and syntax highlighting. Example https://speckyboy.com/snippets-beautiful-syntax-highlighting/

Image: Custom CSS Subsection

![2n7yiBkwvHzUgCXGfBdCu8](https://user-images.githubusercontent.com/37937507/143924805-8eb2e640-566c-48cb-b746-d9d16c26e33b.png)

Image: Custom Editor (Standalone browser window)

![pNrPA82KHXUGD6KWv4DmMZ](https://user-images.githubusercontent.com/37937507/143924907-c4eec154-1da2-4821-a2b2-07608008b285.png)

**Comment by echo6ix** [comment]
> I don't think stash-box integration belongs in Services. Services implies stuff provided by the stash server (ie DLNA). I would say it better belongs in Web Scraping, though that title probably needs generalising as a result.

@WithoutPants Agreed re: Stashbox and scrapers all categorized in the same top-level/primary section. The web scrapers and Stashbox both provide metadata to the users content. Therefore, how about we name the top-level section containing them `Metadata providers`. Kodi calls their add-ons section of metadata scrapers `Information providers`, but I'm partial to just `Metadata providers` personally.

**Comment by echo6ix** [comment]
Library subsection concept image
![library](https://user-images.githubusercontent.com/37937507/143947759-f0054794-7514-495d-9e79-93b18a9fcaac.png)

- Experimented with icons adjacent to the primary (top-level) headings
- I tried to follow some (unofficial) card guidelines for the settings, notably:
  - sub-items are indented and separated using a divider (the divider is also left indented). The opacity and text of the sub-items is slightly less than their parent opacity.

**Comment by echo6ix** [comment]
Jobs queue concepts (I renamed it to tasks queue for consistency since they're referred to as tasks and not jobs throughout the rest of Stash.

Image: Tasks queue is inactive

![tasks queue inactive](https://user-images.githubusercontent.com/37937507/143954394-fcd3b64e-81f5-439c-ac3c-7a03b8b7d174.png)
---------------
Image: Tasks queue active, scanning in process
- The filenames being scanned gets clipped with ellipses to fit the card if they're too long. I used ellipses in the middle of the filename instead of the end - it lets you get a sense of the path and the filename for filenames that will get clipped.
- Stop task button placed on the right to be consistent with the rest of the settings card aesthetic

![tasks queue scanning](https://user-images.githubusercontent.com/37937507/143954468-76f1216b-0ac5-4411-b513-85d8ce6e52e9.png)
---------------
Image: Tasks queue active, auto tagging in process
- Auto tagging progress bar removed. I think it is unnecessary if we show a numerical percentage indicator, which also takes much less space and provides the user the exact same information

![tasks queue autotagging](https://user-images.githubusercontent.com/37937507/143954501-2dc8dbb8-9546-4a92-b9ff-f96238e6fb8c.png)

**Comment by echo6ix** [comment]
Here's a version of the tasks queue that I think incorporates the old layout with new layout

The expand/collapse caret hides all the advanced scanning options.

![nqj5aHJVusF9FQY9LY1Qrj](https://user-images.githubusercontent.com/37937507/143957996-fa992163-e0f2-4d34-91be-12dfdb1da8b6.png)

**Comment by kermieisinthehouse** [comment]
> ![nqj5aHJVusF9FQY9LY1Qrj](https://user-images.githubusercontent.com/37937507/143958318-19d8f03d-7dc9-4e5a-9b72-558a00093f57.png)

This last one is perfect. It's only missing 'Save as default' for the options, and it's done.

**Comment by echo6ix** [comment]
> This last one is perfect. It's only missing 'Save as default' for the options, and it's done.

How's that?

![nqj5aHJVusF9FQY9LY1Qrj](https://user-images.githubusercontent.com/37937507/143959535-644e2146-93f9-4b5d-8fcd-2d78180e43db.png)

**Comment by kermieisinthehouse** [comment]
We don't use switches anywhere else on Stash, but I don't think switching checkboxes to switches across our UI would be so bad. Bootstrap already has some.

I think this design is ideal, especially the collapsible options. We need to make a few more collapsible options across our settings to really simplify it.

**Comment by 7dJx1qP** [comment]
The "Default scan options" button wording looks off to me. I think "Set as default", "Save as default", "Save as default scan options", or "Make default" makes it clearer. Looking at all the buttons in Chrome settings, they all start with a verb when they are buttons that do something.

The task buttons (scan, identify, auto tag) should also probably begin with "Start" as well. Personally, I'm used to the current labels, but I can see new users getting the same feeling about the wording as I did about "Default scan options".

**Comment by echo6ix** [comment]
I am not sure if using switches would be appropriate as a way to select items in grid view. For that section I'd leave checkboxes, and use the switches in the Settings pages, provided the Settings page instantaneously saves changes the user makes. I made
the concepts in the assumption that changes would be auto saved, but should have been clarified that.

>Switches should never require users to press a button to apply the settings because a switch is a toggle button. When you require users to press a submit button, you confuse them because it’s not what they expect.

> You should only use switches on settings that need to take effect instantaneously. If a setting requires a button press before it can take effect, you should use a checkbox instead.

> A checkbox does not apply the settings at an instant like a switch. Instead, it’s has a submit button and takes effect after the user presses it. Requiring a button press allows users to review their settings before they commit. This requirement helps prevent accidental activation errors.

Source: https://uxmovement.com/buttons/when-to-use-a-switch-or-checkbox/

**Comment by WithoutPants** [comment]
I agree that if we move to switches then we should auto-save for everything.

**Comment by jeremymeyers** [comment]
> > This last one is perfect. It's only missing 'Save as default' for the options, and it's done.
> 
> How's that?
> 
> ![nqj5aHJVusF9FQY9LY1Qrj](https://user-images.githubusercontent.com/37937507/143959535-644e2146-93f9-4b5d-8fcd-2d78180e43db.png)
> 
> 

love this and would just suggest that the task queue have both a noticable visual differentiation (as it is showing actions taking place rather than options being set) as well as a notice that the actions will continue even if you navigate to a different page.

**Comment by 7dJx1qP** [comment]
Will the tasks queue still show the output of the current running task? I like seeing the task output for the currently running task, but it can be hard to cancel other tasks, because they're constantly getting shifted up and down by the current task output. I suggest making the output for the current task collapsible.

**Comment by philpw99** [comment]
Loving this conversation ! Great concept !

**Comment by echo6ix** [comment]
I may have made a mistake with all the text input boxes. Because the aim is to have the settings page instantaneously save changes, I think text input boxes will require an `Edit`/`Change`/`Add` type button that opens a small modal window. I noticed that's how the Chromium based browser do it.

Example:

![nqj5aHJVusF9FQY9LY1Qrj](https://user-images.githubusercontent.com/37937507/143969964-f82417eb-9b83-469a-9173-979fae323700.png)

Clicking apply in the modal window would then apply the change.

**Comment by WithoutPants** [comment]
One alternative would be reset and confirm buttons on the text fields to reset/apply the field value. Also note that for some fields we'll need to group them into a dedicated dialog - for example, the authentication username/password setting should be atomic.

**Comment by echo6ix** [comment]
> Will the tasks queue still show the output of the current running task? I like seeing the task output for the currently running task, but it can be hard to cancel other tasks, because they're constantly getting shifted up and down by the current task output. I suggest making the output for the current task collapsible.

@7dJx1qP Yeah, that's why in my concept wireframe the status of the current task, it's output/progress, and it's cancel button are all on the same line. However, that means the output will need to be truncated for long filepaths. But it's up to the devs if that's how they implement it.

> love this and would just suggest that the task queue have both a noticable visual differentiation (as it is showing actions taking place rather than options being set) as well as a notice that the actions will continue even if you navigate to a different page.

@jeremymeyers Hmm, what would you suggest? Different card color maybe? A more visible icon indicator?

I took inspiration entirely from Chromium based browsers. Their Settings > About pages do not have any user options to change, but only show information. For example in the below image where it says the browser is up-to-date, when it is performing a task (downloading), the icon iirc turns into a download spinner, shows a percentage, and has a different status message. Almost par for what the tasks queue concept does.

![nqj5aHJVusF9FQY9LY1Qrj](https://user-images.githubusercontent.com/37937507/143970910-2e239d23-e735-4b84-b7ef-246a9b96e0c7.png)

**Comment by 7dJx1qP** [comment]
> Yeah, that's why in my concept wireframe the status of the current task, it's output/progress, and it's cancel button are all on the same line.

One line isn't enough for task output. This is what I see when I currently see when I run a scan:
![image](https://user-images.githubusercontent.com/38586902/143972797-67c9b412-4e64-4c1e-9998-856ad2a8b60b.png)
The number of lines is constantly changing and I think it depends on the number of threads that are in use

**Comment by echo6ix** [comment]
Oh, I must have misunderstood then. My Stash scans only use one thread so I've never seen that before.

**Comment by jeremymeyers** [comment]
> > Yeah, that's why in my concept wireframe the status of the current task, it's output/progress, and it's cancel button are all on the same line.
> 
> One line isn't enough for task output. This is what I see when I currently see when I run a scan:
> ![image](https://user-images.githubusercontent.com/38586902/143972797-67c9b412-4e64-4c1e-9998-856ad2a8b60b.png)
> The number of lines is constantly changing and I think it depends on the number of threads that are in use

it could be one progress bar for current task completion progress and a second one on a new line for overall queue completion, though that introduces math and would maybe require some assessment of what is in the queue before starting...

Though also I kind of feel like a queue is maybe not the right framework? It is a list of active and pending actions, each of which can be launched individually. Am I wrong in saying a queue in this case would more often be a "stack a bunch of tasks and then hit a Start button and have them all go"

**Comment by 7dJx1qP** [comment]
No, I think conceptually it is a queue. You can start a bunch of tasks and they are queued up with one running at a time.

The issue I'm pointing out is that if a task that's running has multiple threads, it will log the output of each thread right below the task, and the lines are constantly changing and shifting any other queued tasks below it up and down.

These three screenshots shows task output at different points in time of a scan task (with 7 parallel threads) with two other tasks queued. Sometimes there's no thread output if they aren't doing any work, and you can see the other queued tasks below. Other times the whole area will be filled with thread output because they're all doing some work and you won't see the other tasks unless you scroll down. The number of lines changes many times in fractions of a second.
![2021-11-29 21_36_51-Settings _ Stash - Brave](https://user-images.githubusercontent.com/38586902/143976132-24bbbc37-bd86-483f-97f9-3a1d929b0c35.png)
![2021-11-29 21_36_26-Settings _ Stash - Brave](https://user-images.githubusercontent.com/38586902/143976120-e1211d1d-e6d4-4819-8cdb-06d334b67b0c.png)
![2021-11-29 21_36_39-Settings _ Stash - Brave](https://user-images.githubusercontent.com/38586902/143976137-9fe589d6-fd5f-4e7f-9f8f-11121a0bb02b.png)

I think there's a few ways this could be addressed:
1) Move the thread output to a separate section above or below the list of tasks
2) Make the thread output section collapsible
3) Make the thread output section a fixed height regardless of how many lines of output there are. This would stop the rest of the tasks in the list from constantly shifting up and down

**Comment by jeremymeyers** [comment]
> > Will the tasks queue still show the output of the current running task? I like seeing the task output for the currently running task, but it can be hard to cancel other tasks, because they're constantly getting shifted up and down by the current task output. I suggest making the output for the current task collapsible.
> 
> @7dJx1qP Yeah, that's why in my concept wireframe the status of the current task, it's output/progress, and it's cancel button are all on the same line. However, that means the output will need to be truncated for long filepaths. But it's up to the devs if that's how they implement it.
> 
> > love this and would just suggest that the task queue have both a noticable visual differentiation (as it is showing actions taking place rather than options being set) as well as a notice that the actions will continue even if you navigate to a different page.
> 
> @jeremymeyers Hmm, what would you suggest? Different card color maybe? A more visible icon indicator?
> 
> I took inspiration entirely from Chromium based browsers. Their Settings > About pages do not have any user options to change, but only show information. For example in the below image where it says the browser is up-to-date, when it is performing a task (downloading), the icon iirc turns into a download spinner, shows a percentage, and has a different status message. Almost par for what the tasks queue concept does.
> 
> ![nqj5aHJVusF9FQY9LY1Qrj](https://user-images.githubusercontent.com/37937507/143970910-2e239d23-e735-4b84-b7ef-246a9b96e0c7.png)

not to muddy things up here. I get what you are saying about the Edge update section and would just suggest that that is fulfilling a different need (i.e. an update of the entire underlying platform one is using, which would require a restart and is kind of a binary process (either its done updating or its not and no more granular detail is provided) as contrasted by our use case which is "show the current status of tasks the user has selected as well as how much other stuff remains to be done"

Yeah i would put it in a different color or maybe even make it fixed position on top so that the scrolling starts right above "tasks queue" in your wireframe

**Comment by echo6ix** [comment]
I think nitty gritty details like this are gonna come down to user preference. 🤷

> Yeah i would put it in a different color or maybe even make it fixed position on top so that the scrolling starts right above "tasks queue" in your wireframe

Fixing the position might work well on large viewports, but laptop screens and mobile viewports it might be a bad idea. Not sure.

Since the Tasks Queue is the top element in the column it would be pretty easy to use a pseudo-class (eg. `:first-of-type`) in custom css to change the color of the card to whatever you think is appropriate.

I'm currently looking at my TrueNAS dashboard to see if there's anything similar. It does have a Task Manager of sorts (itemized tasks progress with pct% indicator), but it is utilized as toast content when you click the task icon in the nav bar.  Fwiw I think something like that is being added to the nav bar, just not sure how interactive it is going to be.

**Comment by QxxxGit** [comment]
> ### Custom CSS subsection
> 
> To reduce clutter on the main settings page and to increase the usability of the Custom CSS input box, herein referred to as the `CSS Editor`, add an external launch icon that opens a new browser window specifically for editing custom CSS.
>
> [...]

IMO, I think it's fine to keep the textarea as-is until a long term solution for easily installing themes comes to fruition. Most people are going to edit CSS in an application of their choice anyways i.e. VS Code. Would just be unnecessary work, IMO.

Other than that, I think switches are the way to go. I think they'll look better/modernized than checkboxes on mobile too. I also really like the new layout echo6ix conceptualized. It's clean.

**Comment by echo6ix** [comment]
It's more about design consistency tbh. Text input forms, especially mutlivalue ones should get a modal window, otherwise i think we're gonna need a `save` or `apply` button after every text input item since the global save button is being deprecated. I mean, the CSS Editor would be just another window, it doesn't have to be decked out with line numbers, alternate line coloring, and syntax highlighting. It can be pretty basic for the first iterations for the reasons you suggested.

**Comment by echo6ix** [comment]
Here's another mistake I think I made in my concepts. In every image showing a caret or chevron meant to expand/collapse icon, I incorrectly used an up caret/chevron to indicate the content was in a collapsed state.

### Wrong usage of expand/collapse icons
- ▾ Expanded state
- ▴ Collapsed state 

### Correct usage of expand/collapse icons
- When icon is located _left_ of text
  - ▾ Expanded state
  - ▸ Collapsed state
- When icon is located _right_ of text
  - Expanded state ▾
  - Collapsed state ◂

**Comment by jeremymeyers** [comment]
> I think nitty gritty details like this are gonna come down to user preference. 🤷
> 
> > Yeah i would put it in a different color or maybe even make it fixed position on top so that the scrolling starts right above "tasks queue" in your wireframe
> 
> Fixing the position might work well on large viewports, but laptop screens and mobile viewports it might be a bad idea. Not sure.
> 
> Since the Tasks Queue is the top element in the column it would be pretty easy to use a pseudo-class (eg. `:first-of-type`) in custom css to change the color of the card to whatever you think is appropriate.
> 
> I'm currently looking at my TrueNAS dashboard to see if there's anything similar. It does have a Task Manager of sorts (itemized tasks progress with pct% indicator), but it is utilized as toast content when you click the task icon in the nav bar. Fwiw I think something like that is being added to the nav bar, just not sure how interactive it is going to be.

I mean this is outside the scope of this discussion but at some point possibly adding a footer bar to communicate information might be valuable?

**Comment by kermieisinthehouse** [comment]
Closing as the settings page has been refactored, and released in 0.12.0. Great work on the design!
